ext/opcache/jit/zend_jit_arm64.dasc

```diff
@@ -483,10 +483,100 @@ static int logical_immediate_p (uint64_t value, uint32_t reg_size)
 ||	}
 |.endmacro
 
+#ifdef __APPLE__
+|.macro SAVE_REGS
+|	sub sp, sp, #(16*21)
+|
+|	stp d30, d31, [sp]
+|	stp d28, d29, [sp, #16]
+|	stp d26, d27, [sp, #(16*2)]
+|	stp d24, d25, [sp, #(16*3)]
+|	stp d22, d23, [sp, #(16*4)]
+|	stp d20, d21, [sp, #(16*5)]
+|	stp d18, d19, [sp, #(16*6)]
+|	stp d16, d17, [sp, #(16*7)]
+|	// Callee-saved registers, d8 to d15, can be omitted.
+|	stp d6, d7, [sp, #(16*8)]
+|	stp d4, d5, [sp, #(16*9)]
+|	stp d2, d3, [sp, #(16*10)]
+|	stp d0, d1, [sp, #(16*11)]
+|
+|	stp x29, x30, [sp, #(16*12)]
+|	// Callee-saved registers, x19 to x28, can be omitted.
+|	// Temporary registers TMP1 to TMP3, i.e. x15 to 17, can be omitted.
+|	stp x14, x18, [sp, #(16*13)]
+|	stp x12, x13, [sp, #(16*14)]
+|	stp x10, x11, [sp, #(16*15)]
+|	stp x8, x9, [sp, #(16*16)]
+|	stp x6, x7, [sp, #(16*17)]
+|	stp x4, x5, [sp, #(16*18)]
+|	stp x2, x3, [sp, #(16*19)]
+|	stp x0, x1, [sp, #(16*20)]
+|.endmacro
+
+|.macro RESTORE_REGS
+|	ldp d30, d31, [sp]
+|	ldp d28, d29, [sp, #16]
+|	ldp d26, d27, [sp, #(16*2)]
+|	ldp d24, d25, [sp, #(16*3)]
+|	ldp d22, d23, [sp, #(16*4)]
+|	ldp d20, d21, [sp, #(16*5)]
+|	ldp d18, d19, [sp, #(16*6)]
+|	ldp d16, d17, [sp, #(16*7)]
+|	// Callee-saved registers, d8 to d15, can be omitted.
+|	ldp d6, d7, [sp, #(16*8)]
+|	ldp d4, d5, [sp, #(16*9)]
+|	ldp d2, d3, [sp, #(16*10)]
+|	ldp d0, d1, [sp, #(16*11)]
+|
+|	ldp x29, x30, [sp, #(16*12)]
+|	// Callee-saved registers, x19 to x28, can be omitted.
+|	// Temporary registers TMP1 to TMP3, i.e. x15 to 17, can be omitted.
+|	ldp x14, x18, [sp, #(16*13)]
+|	ldp x12, x13, [sp, #(16*14)]
+|	ldp x10, x11, [sp, #(16*15)]
+|	ldp x8, x9, [sp, #(16*16)]
+|	ldp x6, x7, [sp, #(16*17)]
+|	ldp x4, x5, [sp, #(16*18)]
+|	ldp x2, x3, [sp, #(16*19)]
+|	ldp x0, x1, [sp, #(16*20)]
+|
+|	add sp, sp, #(16*21)
+|.endmacro
+#endif
+
 |.macro LOAD_TSRM_CACHE, reg
+||#ifdef __APPLE__
+|	// TODO: Get a better name for "tsrm_ls_cache_tcb_offset"?
+|	// TODO: In Linux only TMP3 is used particularly for ZTS, whereas TMP1 and TMP2 are also used in macOS.
+|	LOAD_ADDR TMP3, tsrm_ls_cache_tcb_offset
+|	.long 0xd53bd06f // TODO: hard-coded: mrs TMP1, tpidrro_el0
+|	and TMP1, TMP1, #0xfffffffffffffff8
+|	ldr TMP2, [TMP3, #0x8]
+|	ldr TMP1, [TMP1, TMP2, lsl #3]
+|	cbz TMP1, >4     // TODO: There is conflict for local labels. A number of test cases would fail.
+|	ldr TMP2, [TMP3, #0x10]
+|	ldr reg, [TMP1, TMP2]
+|	b >7
+|4:
```

Comment:
Discussion:
The slow path is common code and it would be good to put it in some global stub, however, we have to jump back from global stub.
As per "cold code", this would increase the code size since this piece of code would be generated for each invocation of macro "LOAD_TSRM_CACHE".

Another solution might be that, we access variable `tsrm_ls_cache` just after the thread is created, and in this way the slow path can be avoided. Yes. This would be correct with the premise that **slow path** is executed only once for each thread.

Reply:
It seems that it works in my local testing environment with the following patch. All test cases can pass. But I'm not sure if this is sound.
```
diff --git a/TSRM/TSRM.c b/TSRM/TSRM.c
index 18cf0d163c..c2907dee56 100644
--- a/TSRM/TSRM.c
+++ b/TSRM/TSRM.c
@@ -745,6 +745,7 @@ TSRM_API size_t tsrm_get_ls_cache_tcb_offset(void)
        size_t ret;

 # ifdef __APPLE__
+       ret = _tsrm_ls_cache;
        // In macOS, "ret" holds the address of one TLVDecriptor, where the first
        // member is the address of function tlv_get_addr().
        asm("adrp %0, #__tsrm_ls_cache@TLVPPAGE\n\t"
diff --git a/ext/opcache/jit/zend_jit_arm64.dasc b/ext/opcache/jit/zend_jit_arm64.dasc
index 0fc88a5090..615c08a749 100644
--- a/ext/opcache/jit/zend_jit_arm64.dasc
+++ b/ext/opcache/jit/zend_jit_arm64.dasc
@@ -554,24 +554,8 @@ static int logical_immediate_p (uint64_t value, uint32_t reg_size)
 |      and TMP1, TMP1, #0xfffffffffffffff8
 |      ldr TMP2, [TMP3, #0x8]
 |      ldr TMP1, [TMP1, TMP2, lsl #3]
-|      cbz TMP1, >4     // TODO: There is conflict for local labels. A number of test cases would fail.
 |      ldr TMP2, [TMP3, #0x10]
 |      ldr reg, [TMP1, TMP2]
-|      b >7
-|4:
-|      // Slow path. Invoke tlv_get_addr(). TODO: Put it in cold code or global stub?
-|      stp x29, x30, [sp, #-16]!
-|      mov x29, sp
-|      SAVE_REGS
-|      mov x0, TMP3
-|      ldr TMP1, [TMP3]
-|      blr TMP1
-|      mov TMP3, x0
-|      RESTORE_REGS     // TMP3 is not in this list
-|      ldr reg, [TMP3]
-|      mov sp, x29
-|      ldp, x29, x30, [sp], #16
-|7:
 ||#else
 |      .long 0xd53bd051 // TODO: hard-coded: mrs TMP3, tpidr_el0
 ||     ZEND_ASSERT(tsrm_ls_cache_tcb_offset <= LDR_STR_PIMM64);
```